src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1745,9 +1759,25 @@ protected void AfterMembersChecks(BindingDiagnosticBag diagnostics)
                 }
             }
 
-            if (IsFileLocal && (object?)ContainingType != null)
+            if (IsFileLocal)
             {
-                diagnostics.Add(ErrorCode.ERR_FileTypeNested, location, this);
+                // A well-behaved file-local type only has declarations in one syntax tree.
+                // There may be multiple syntax trees across declarations in error scenarios,
+                // but we're not interested in handling that for the purposes of producing this error.
+                var tree = declaration.Declarations[0].SyntaxReference.SyntaxTree;
+                // TODO2: we probably don't want to linearly search the syntax trees.
+                foreach (var otherTree in DeclaringCompilation.SyntaxTrees)
+                {
+                    if ((object)tree != otherTree && tree.FilePath == otherTree.FilePath)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_FileTypeNonUniquePath, location, this, tree.FilePath);
+                    }
+                }
```

Comment:
Would it make sense to check for unique paths during emit, where we could generate a dictionary (mapping syntax tree to file path) from the file-local types? (We have access to the complete set of top-level types - see `PEModuleBuilder.GetTopLevelSourceTypeDefinitions()` for instance.)